src/jailer/src/env.rs

```diff
@@ -197,37 +200,53 @@ impl Env {
     }
 
     fn exec_into_new_pid_ns(&mut self, chroot_exec_file: PathBuf) -> Result<()> {
-        // Unshare into a new PID namespace.
-        // The current process will not be moved into the newly created namespace, but its first
-        // child will assume the role of init(1) in the new namespace.
-        // The call is safe because we're invoking a C library function with valid parameters.
-        SyscallReturnCode(unsafe { libc::unshare(libc::CLONE_NEWPID) })
-            .into_empty_result()
-            .map_err(Error::UnshareNewPID)?;
-
         // Compute jailer's total CPU time up to the current time.
         self.jailer_cpu_time_us =
             utils::time::get_time_us(utils::time::ClockType::ProcessCpu) - self.start_time_cpu_us;
 
-        // Duplicate the current process. The child process will belong to the previously created
-        // PID namespace.
-        // TODO: replace the `unshare()` + `fork()` combo with `clone()` if we ever need to
-        //  squeeze every bit of start-up latency we can get
-        let pid = unsafe { libc::fork() };
-        match pid {
-            0 => {
-                // Reset process start time.
-                self.start_time_cpu_us = 0;
-
-                Err(Error::Exec(self.exec_command(chroot_exec_file)))
-            }
-            child_pid => {
-                // Save the PID of the process running the exec file provided
-                // inside <chroot_exec_file>.pid file.
-                self.save_exec_file_pid(child_pid, chroot_exec_file)?;
-                unsafe { libc::exit(0) }
-            }
+        // This is a wrapper for our child callback to the clone libc function.
+        // We will use clone's arg parameter as the actual callback and call it from here.
+        extern "C" fn child_fn(arg: *mut libc::c_void) -> libc::c_int {
+            let callback_ref: *mut *mut dyn FnOnce() = arg as *mut *mut dyn std::ops::FnOnce();
+            let callback = unsafe { Box::from_raw(*callback_ref) };
+
+            callback();
+            0
         }
+
+        let child_callback = || {
+            // Reset process start time.
+            self.start_time_cpu_us = 0;
+
+            self.exec_command(chroot_exec_file.clone());
+        };
+
+        let child_callback_box: Box<dyn FnOnce()> = Box::new(child_callback);
```

Comment:
So this will cause a segfault too, we have to let rust know it is a FnOnce instance.

Reply:
Hmmm this is concerning because it would compile and then we would get an error at runtime. It might be easy to introduce bugs this way. 

Anyway, this problem, the double pointers problem, and the nested callbacks problem are all caused by the fact that we use the closure. Another option would be to capture the variables ourselves. Something like:

```
        struct Capture<'a> {
            env: &'a mut Env,
            chroot_exec_file: PathBuf,
        }

        let capture = Capture {
            env: self,
            chroot_exec_file,
        };
        let capture_ptr = Box::into_raw(Box::new(capture));
```

and keep only `on_clone`:

```
        extern "C" fn on_clone(arg: *mut libc::c_void) -> libc::c_int {
            let capture = unsafe { Box::<Capture>::from_raw(arg as _) };

            capture.env.exec_command(capture.chroot_exec_file.clone());

            0
        }
```

Not sure which option is best.

Reply:
Also one more question comes to mind here. Sorry for insisting, but I am not very familiar with `clone()` and I would like to understand all the implications of this PR.

Whether we use a closure or we create a structure for passing arguments to the `on_clone()` callback, we end up passing a reference to self (For example the `env: &'a Env` in the previous comment). AFAIU If the `Env` struct was allocated on the parent's heap, we will find it in the child's process memory as well. But if the `Env` struct was allocated on the parent's stack, what will happen when `on_clone()` is executed ? What will the reference that we have in `Capture` point to ?

Reply:
So one very important thing you have to understand about clone - in our case (the flags I used) are really similar to the fork that was here before the commit. You can actually think about this change as a fork with the new-pid-ns flag on.
You can [look](https://github.com/lattera/glibc/blob/416bf844227d37b043b16be28c9523eeaecd3de3/nptl/sysdeps/unix/sysv/linux/x86_64/fork.c) on fork libc's source (that actually uses the fork syscall) and see it is exactly the same (except some thread flags that are really not important for us. 

The clone will actually copy all the memory to a new process just like fork does. Also, I think the capture is much more elegant and will be easier to understand than capturing all of them ourselves. There are also some disadvantages on capturing them ourselves BTW. I think all this part is ok and we can move on.

Reply:
Ok, can you help me understand more in-depth these details ? Again, I don't know exactly how `clone()` works and I think these are crucial details, especially in the jailer where it can have security implications. I also started from the assumption that the way it is called, `clone()` should perform like `fork()`. But is this assumption true ? There are some corner cases that I can think of where I don't know how it would behave. So expanding on my previous question. What if the parent stack is 10k bytes and the `Env` or a captured variable is at the end of the parent's stack. Will it be copied ? If we used the original `clone()` syscall and we provide a null pointer as stack, my understanding is that indeed everything would have been copied identically. But what happens now since we allocate a stack of size 8192 bytes for the child ? What are the implications of doing this ?

> You can look on fork libc's source (that actually uses the clone syscall) and see it is exactly the same (except some thread flags that are really not important for us.

It's also providing a NULL stack, right ? 

> Also, I think the capture is much more elegant and will be easier to understand than capturing all of them ourselves. 

I agree. The capture looks better cosmetically.

Reply:
Ok, we shall look on the whole flow :)

1. We call libc's clone from the jailer.
2. Libc is [translating](https://code.woboq.org/userspace/glibc/sysdeps/unix/sysv/linux/x86_64/clone.S.html
) to system call ABI and dispatching a clone syscall.
It is important to notice that the clone syscall gives different execution pathes acording to the return value (very similar to fork's behavior) and that libc translates it to a callback API for the user.
3. In the [kernel](https://code.woboq.org/linux/linux/kernel/fork.c.html#2327
), clone is defined to directly call _do_fork kernel function
4. Some unimportant checks for us, then calling [copy_process](https://code.woboq.org/linux/linux/kernel/fork.c.html#2226
)
5. A call to copy_mm is [made](https://code.woboq.org/linux/linux/kernel/fork.c.html#1916
), with the new process and the clone flags
6. We will take the [flow](https://code.woboq.org/linux/linux/kernel/fork.c.html#1374
) of CLONE_VM is not in the flags, which results in a deep copy of the mm of the original process
7. You can [see](https://code.woboq.org/linux/linux/kernel/fork.c.html#dup_mm
) that dup_mm actually allocates and memcpy the memory of the mm.

All of this will result in a FULL copy of the original process memory, including the original process stack :)

Reply:
Btw, the actual sp setting is done [here](https://code.woboq.org/linux/linux/arch/x86/kernel/process_64.c.html#420) inside copy_thread_tls which is called [here](https://code.woboq.org/linux/linux/kernel/fork.c.html#1925) inside copy_process.

Also, I made a boilerplate testing code that:
1. writes a value to its stack
2. call libc's clone
3. original father process exits
4. child process reads the value (after death) 

and everything is ok because as we said, this memory is copied. Thats it, I hope I shed some light for you. If anything is ok I think we finished everything except the return value of clone (@luminitavoicu ) and we just need to dispatch the tests in the buildkite and finally moving on :100:

Reply:
Thanks for the detailed explanations and links to kernel code. I will try to follow them in order to understand what the kernel does exactly. It will take some time. Anyway, yes, probably we're ok here. Just want to double check.

Reply:
I looked on the kernel code and did some experiments. You're right, when calling clone the entire address space of the parent is copied and the parent's stack variables can be accessed at the same address as in the parent. The only problem is that as the child's stack grows it will eventually overwrite the previously allocated variables. For example in this sample code:

```
const PROCESS_STACK_SIZE: usize = 32;

fn main() {
    extern "C" fn on_clone(arg: *mut libc::c_void) -> libc::c_int {
        let data = arg as *mut u8;
        // Here (*data) == 37
        let stack_var_1 = 1u64;
        let stack_var_2 = 2u64;
        let stack_var_3 = 3u64;
        let stack_var_4 = 4u64;
        // Here (*data) == 4

        0
    }

    let mut data = 37u8;
    let data_ptr = &mut data as *mut u8;
    let arg: *mut libc::c_void = data_ptr as _;

    let mut stack = [0u8; PROCESS_STACK_SIZE];
    let stack_top_ptr = stack.last_mut().unwrap() as *mut u8;

    let child_pid = unsafe {
        libc::clone(
            on_clone,
            stack_top_ptr as _,
            libc::SIGCHLD | libc::CLONE_NEWPID,
            arg,
        )
    };

    if child_pid < 0 {
        println!("clone error: {}", std::io::Error::last_os_error());
    } else {
        println!("child_pid: {}", child_pid);
    }
}
```

You can see that we allocate `data = 37` and we pass it to the `on_clone()` callback. When we enter `on_call()` data is indeed 37, but as we allocate variables on the stack it is eventually overwritten and it becomes `4`. Not sure if this represents any reason for concern. I would just like to spend some time thinking about it. @firecracker-microvm/compute-capsule what do you think ?

Reply:
After spending some more time researching about this I think that the example above shows that it's hard to predict exactly where a variable will be allocated on the stack. That's because of compiler optimizations and the design of the code. Ideally we want the child's stack (`let mut stack = [0u8; PROCESS_STACK_SIZE];`) to be allocated at the bottom of the paren't stack, but we can't guarantee that with the current approach. Sometimes other variables might be allocated below it in which case we can end up overwriting them as the child's stack grows. This can be problematic if we actually need them in the child. I can think of 3 ways of allocating the child's stack:

1. As a slice allocated on the paren't stack. This is the current approach in the PR, and it doesn't look safe because of the reasons listed above.

2. On the heap with mmap. I'm not sure if this is ok since we don't know exactly where this memory is going to be mapped and if it's mapped really low in the address space, there might be little space for the stack and heap to grow. I'm not completely sure about this reasoning, but that's how it seems like. 

3. Using the clone syscall directly and providing a NULL stack pointer to it. I assume and hope that this option places the child's stack right at the bottom of the parent's stack which would be perfect. If this assumption is true, this seems like the safest option to me and I'm leaning towards it.

Please let me know if I'm missing anything.

Reply:
So:
2. If we will use mmap we can choose exactly the place of our mapping and place it where we would like - this will require extra logic and result in less clean code and more syscalls.
3. So this is not what will happen - if we will provide a NULL stack pointer, we will just continue with the original stack - and that's ok because as we talked, the VM_CLONE flag is not in use, and we will have a new replicated memory space - and this is exactly what fork do !

I will implement the raw syscall option, because we both understand this is the right option. I just want you to understand that there is no stack allocated in this case, because all the memory is copied and the virtual addresses will remain the same :)

Reply:
> 3. So this is not what will happen - if we will provide a NULL stack pointer, we will just continue with the original stack - and that's ok because as we talked, the VM_CLONE flag is not in use, and we will have a new replicated memory space - and this is exactly what fork do !

> I will implement the raw syscall option, because we both understand this is the right option. I just want you to understand that there is no stack allocated in this case, because all the memory is copied and the virtual addresses will remain the same :)

I see. It makes sense and indeed it seems the right option to me.